### ISSUE TITLE
Diagonal space in Urdu Nastaleeq font

### ISSUE THREAD

**Issue by azeemdin** [tesseract-ocr/tesseract#2407]
Hi,
I am trying to train Tesseract4 with (ocrd-train) Makefile except box file generation. Box file(s) are generated using below command

`text2image --text "/traintext.txt"  --outputbase "/traintext"  --fontconfig_tmpdir "/fontconfig"  --fonts_dir "/usr/share/fonts" --font "Jameel Noori Nastaleeq" --leading 32`

I have used (25000) iterations by updating Makefile. I have used following command for generating traindata

`sudo make training MODEL_NAME=urd1 START_MODEL=urd TESSDATA_REPO=_best WORD_LIST=urd1.worldlist.clean
`
Start model is existing Urdu model from _best repository.
(sample files are attached)
[nastaleeq_test.zip](https://github.com/tesseract-ocr/tesseract/files/3113397/nastaleeq_test.zip)

"Jameel Noori Nastqleeq" font is ligature based. After successful training when I try to use following code to perform OCR on below image, I am facing an issue related to space between few words

``
public static void main(String[] args) {
		File imageFile = new File("testing.png");
		ITesseract instance = new Tesseract(); 
		instance.setDatapath("tessdata");
		try {
			instance.setLanguage("urd1")
			String result = instance.doOCR(imageFile);
			System.out.println(result);
		} catch (TesseractException e) {
			System.err.println(e.getMessage());
		}
	}
``
![testing3](https://user-images.githubusercontent.com/7873519/56678798-3b8d7a80-66cc-11e9-99f8-db5efed4ea02.png)
OCR result:
`ہفتوںکی منصوبہ بندی حقیقتکا روپ دھار نےلگی
`
This is same line which is used during training. OCR Output is fine except the missing space between few words i.e.
![result](https://user-images.githubusercontent.com/7873519/56678839-4ea04a80-66cc-11e9-8375-0381996255ba.png)
I believe it is because space in Urdu Nastaleeq (with kerning) writing style is diagonal instead of vertical
![diagonalspace](https://user-images.githubusercontent.com/7873519/56678853-565fef00-66cc-11e9-94d9-55bccee17f0f.png)

Is there any possibility to support diagonal space in Tesseract4?

**Comment by Shreeshrii** [comment]
1. Is this the only accuracy issue that you are facing with the finetuned model?
2. Have you tested the model on images that it was not trained on?

**Comment by azeemdin** [comment]
Thanks Shreeshrii for reply.

> 1. Is this the only accuracy issue that you are facing with the finetuned model?
There are other issues related to identification of glyphs where characters overlap each other. 
<img width="164" alt="testing3" src="https://user-images.githubusercontent.com/7873519/56846834-a7dec880-68dc-11e9-8930-1e36d46e2c9c.png">
Result is:
`مجھے دس سال پہلےسیکی تاغش نہیں لینی تھی`
where as correct one is
`مجھے دس سال پہلے کسی کی تلاشی نہیں لینی تھی`
> 2. Have you tested the model on images that it was not trained on?
Yes, tested few images and result is 60% to 80%. Which I think can be improved with more training text and iterations.

**Comment by amitdo** [comment]
Possible partial workaround:

Post process the results and add space after final form of letters.

**Comment by azeemdin** [comment]
> Possible partial workaround:
> 
> Post process the results and add space after final form of letters.

Thanks Amit, most of the available solutions (as per my knowledge) are doing the same post processing in different ways but post processing algorithm itself is not that simple due to the complexity of ligatures in Urdu. By the way we are trying to implement a suitable post processing algorithm to achieve maximum possible correctness.

**Comment by ahsan-raazaa** [comment]
Azeemdin,
I am trying to build urdu ocr but the default tesseract OCR  (urd.traineddata) are not giving promising results.
I like to know how well your ocr perform on different Urdu fonts.

**Comment by azeemdin** [comment]
Salam Ahsan,
Unfortunately due to some other commitments I am unable to proceed working
on it, it is in my wish list but don't know when I am going to find proper
time for it.

Default traindata works fine with "*Naskh*" fonts. The difficult part is
with "*Nastaleeq*" fonts (as mentioned earlier) due to to it's cursive
nature and Kerning. I believe (not sure) we can train tesseract with proper
training data with "Nastaleeq" fonts but that requires hell amount of time
to create images with relevant box files (these can be auto generated but
auto generated files didn't work with my test cases, so I think auto
generated files might need to be reviewed/updated manually before feeding
in tesseract engine).

Best result so far I have seen are with Google vision API and by using it I
have noticed (again not sure) that they are doing post processing as they
have huge infrastructure as well as large amount of data. It seems they are
doing some context analysis in post processing before generating final
results. I tried some post processing algorithms with tesseract engine
(using my own traindata file created using Nastaleeq fonts) output, without
any positive results.

One of the OCR application using tesseract engine 3.+ was developed by some
fellows in UET (Lahore). If you have an urgent requirement you can check
their product. It produces good results (with preconditions).

www.urduocr.net

--------------------------------------------
Muhammad Azeem-ud-din

**Comment by ahsan-raazaa** [comment]
Walaikum salam
Thank you so much for replying

**Comment by sameearif88** [comment]
Aoa Azeemudin,

I have a question regarding training process of tesseract for Urdu. Can you please provide me your email id so i can email you?

Thank you
